packages/runtime/container-runtime/src/scheduleManager.ts

```diff
@@ -27,63 +61,93 @@ type IRuntimeMessageMetadata = undefined | {
  */
 export class ScheduleManager {
     private readonly deltaScheduler: DeltaScheduler;
-    private batchClientId: string | undefined;
-    private hitError = false;
+    private state: BatchProcessingState = "None";
 
     constructor(
         private readonly deltaManager: IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>,
         private readonly emitter: EventEmitter,
+        readonly clientId: () => string | undefined,
         private readonly logger: ITelemetryLogger,
     ) {
         this.deltaScheduler = new DeltaScheduler(
             this.deltaManager,
             ChildLogger.create(this.logger, "DeltaScheduler"),
         );
-        void new ScheduleManagerCore(deltaManager, logger);
+        void new ScheduleManagerCore(deltaManager, clientId, logger);
     }
 
     public beforeOpProcessing(message: ISequencedDocumentMessage) {
-        if (this.batchClientId !== message.clientId) {
-            assert(this.batchClientId === undefined,
-                0x2a2 /* "Batch is interrupted by other client op. Should be caught by trackPending()" */);
-
-            // This could be the beginning of a new batch or an individual message.
-            this.emitter.emit("batchBegin", message);
-            this.deltaScheduler.batchBegin(message);
-
-            const batch = (message?.metadata as IRuntimeMessageMetadata)?.batch;
-            if (batch) {
-                this.batchClientId = message.clientId;
-            } else {
-                this.batchClientId = undefined;
+        const messageState = messageBatchState(message);
+        switch (this.state) {
+            case "Batch":
+                assert(messageState !== "StartOfBatch", "Batch was interrupted by another batch");
+                // We're already processing a batch, do nothing
+                return;
+
+            case "Individual":
+                assert(messageState !== "StartOfBatch", "Batch was interrupted by another batch");
```

Comment:
>another batch [](http://example.com/codeflow?start=82&length=13)

"by an individual op" in this case ?

Reply:
Technically yes, but the op needs to have `batch === true` so that implies the start of a new batch.

Reply:
yes, I saw it later :)